File: C#/LocalSearchOptimization/Solvers/ParallelMultistart.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Threading;
using System.Threading.Tasks;
using LocalSearchOptimization.Components;
using LocalSearchOptimization.Parameters;

namespace LocalSearchOptimization.Solvers
{
    public class ParallelMultistart : IOptimizationAlgorithm
    {
        private MultistartParameters parameters;

        private ISolution currentSolution;

        private List<SolutionSummary> searchHistory;

        private bool stopFlag = false;

        public ISolution CurrentSolution { get => currentSolution; }

        public List<SolutionSummary> SearchHistory { get => searchHistory; }

        private object thisLock = new object();

        public ParallelMultistart(MultistartParameters parameters)
        {
            this.parameters = parameters;
        }

        public IEnumerable<ISolution> Minimize(ISolution startSolution)
        {
            stopFlag = false;
            OptimizationParameters coreParameters = parameters.Parameters;
            Random random = new Random(parameters.Seed);
            DateTime startedAt = DateTime.Now;
            ISolution bestSolution = startSolution;
            currentSolution = startSolution;
            startSolution.IterationNumber = 0;
            startSolution.TimeInSeconds = 0;
            startSolution.IsCurrentBest = false;
            startSolution.IsFinal = false;
            startSolution.InstanceTag = coreParameters.Name;
            searchHistory = new List<SolutionSummary>();
            List<ISolution> solutions = new List<ISolution>();
            Task<int>[] solvers = new Task<int>[parameters.InstancesNumber];
            for (int i = 0; i < parameters.InstancesNumber; i++)
            {
                OptimizationParameters instanceParameters = coreParameters.Clone();
                instanceParameters.Name = (parameters.InstancesNumber > 1 ? parameters.Name + "|" + i + "|" : "") + coreParameters.Name;
                instanceParameters.Seed = random.Next();
                ISolution instanceStartSolution = parameters.RandomizeStart ? startSolution.Shuffle(instanceParameters.Seed) : startSolution;
                solvers[i] = Task<int>.Factory.StartNew(() => Solve(parameters.OptimizationAlgorithm, instanceParameters, instanceStartSolution, solutions));
            }
            using (ManualResetEventSlim delay = new ManualResetEventSlim(initialState: false))
            {
                while (solvers.Any(x => !x.IsCompleted) || solutions.Count() > 0)
                {
                    delay.Wait(parameters.OutputFrequency);
                    lock (thisLock)
                        if (solutions.Count > 0)
                        {
                            ISolution current = solutions.OrderBy(x => x.CostValue).First();
                            current.TimeInSeconds = (DateTime.Now - startedAt).TotalSeconds;
                            current.IsCurrentBest = false;
                            current.IsFinal = false;
                            searchHistory.AddRange(solutions.Select(x => new SolutionSummary
                            {
                                InstanceTag = x.InstanceTag,
                                OperatorTag = x.OperatorTag,
                                IterationNumber = x.IterationNumber,
                                CostValue = x.CostValue
                            }));
                            if (current.CostValue < bestSolution.CostValue)
                            {
                                current.IsCurrentBest = true;
                                currentSolution = current;
                                bestSolution = current;
                                yield return bestSolution;
                            }
                            else if (parameters.DetailedOutput)
                            {
                                currentSolution = current;
                                yield return current;
                            }
                            solutions.Clear();
                        }
                }
            }
            bestSolution.IterationNumber = solvers.Sum(x => x.Result);
            bestSolution.TimeInSeconds = (DateTime.Now - startedAt).TotalSeconds;
            bestSolution.IsCurrentBest = true;
            bestSolution.IsFinal = true;
            currentSolution = bestSolution;
            yield return currentSolution;
        }

        public void Stop()
        {
            stopFlag = true;
        }

        private int Solve(Type optimizationAlgorithm, OptimizationParameters parameters, ISolution startSolution, List<ISolution> output)
        {
            int iteration = 0;
            IOptimizationAlgorithm solver = (IOptimizationAlgorithm)Activator.CreateInstance(optimizationAlgorithm, new object[] { parameters });
            foreach (ISolution solution in solver.Minimize(startSolution))
            {
                lock (thisLock) output.Add(solution);
                iteration = solution.IterationNumber;
                if (stopFlag) solver.Stop();
            }
            return iteration;
        }
    }
}

File: java/src/localsearchoptimization/solvers/ParallelMultistart.java
package localsearchoptimization.solvers;

import localsearchoptimization.components.OptimizationAlgorithm;
import localsearchoptimization.components.Solution;
import localsearchoptimization.components.SolutionHandler;
import localsearchoptimization.components.SolutionSummary;
import localsearchoptimization.parameters.OptimizationParameters;
import localsearchoptimization.parameters.MultistartParameters;

import java.lang.reflect.InvocationTargetException;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Random;
import java.util.concurrent.RecursiveAction;
import java.util.stream.Collectors;

public class ParallelMultistart implements OptimizationAlgorithm {

    private MultistartParameters parameters;

    private SolutionHandler solutionHandler;

    private Solution currentSolution;

    private ArrayList<SolutionSummary> searchHistory;

    private boolean stopFlag = false;

    private final Object thisLock = new Object();

    public ParallelMultistart(MultistartParameters parameters) {
        this(parameters, null);
    }

    public ParallelMultistart(MultistartParameters parameters, SolutionHandler solutionHandler) {
        this.parameters = parameters;
        this.solutionHandler = solutionHandler;
    }

    @Override
    public Solution minimize(Solution startSolution) {
        stopFlag = false;
        Random random = new Random(parameters.seed);
        long startedAt = System.currentTimeMillis();
        Solution bestSolution = startSolution;
        currentSolution = startSolution;
        startSolution.iterationNumber(0);
        startSolution.elapsedTime(0);
        startSolution.isCurrentBest(false);
        startSolution.isFinal(false);
        startSolution.instanceTag(parameters.name);
        searchHistory = new ArrayList<SolutionSummary>();
        ArrayList<Solution> solutions = new ArrayList<Solution>();
        RecursiveAction[] solvers = new RecursiveAction[parameters.instancesNumber];
        for (int i = 0; i < parameters.instancesNumber; i++) {
            OptimizationParameters instanceParameters = parameters.parameters.Clone();
            instanceParameters.name = parameters.parameters.name + (parameters.instancesNumber > 1 ? ":" + i : "");
            instanceParameters.seed = random.nextInt();
            Solution instanceStartSolution = startSolution.shuffle(instanceParameters.seed);
            solvers[i] = new Solver(parameters.optimizationAlgorithm, instanceParameters, startSolution, solutions);
            solvers[i].fork();
        }

        while (Arrays.stream(solvers).anyMatch(x -> !x.isDone()) || solutions.size() > 0) {
            try {
                Thread.sleep(parameters.outputFrequency);
                synchronized (thisLock) {
                    if (solutions.size() > 0) {
                        Solution current = solutions.stream().sorted((f1, f2) -> Double.compare(f1.cost(), f2.cost())).findFirst().get();
                        current.elapsedTime((System.currentTimeMillis() - startedAt) / 1000.0);
                        current.isCurrentBest(false);
                        current.isFinal(false);
                        solutions.stream().map(x -> new SolutionSummary(x.instanceTag(), x.operatorTag(), x.iterationNumber(), x.cost())).collect(Collectors.toCollection(() -> searchHistory));
                        if (current.cost() < bestSolution.cost()) {
                            current.isCurrentBest(true);
                            bestSolution = current;
                            currentSolution = current;
                            if (solutionHandler != null)
                                solutionHandler.process(currentSolution);
                        } else if (parameters.isDetailedOutput) {
                            currentSolution = current;
                            if (solutionHandler != null)
                                solutionHandler.process(currentSolution);
                        }
                        solutions.clear();
                    }
                }
            } catch (InterruptedException e) {
                e.printStackTrace();
            }
        }
        bestSolution.iterationNumber(Arrays.stream(solvers).mapToInt(x -> ((Solver) x).solver.currentSolution().iterationNumber()).sum());
        bestSolution.elapsedTime((System.currentTimeMillis() - startedAt) / 1000.0);
        bestSolution.isCurrentBest(true);
        bestSolution.isFinal(true);
        currentSolution = bestSolution;
        if (solutionHandler != null)
            solutionHandler.process(currentSolution);
        return currentSolution;
    }

    @Override
    public Solution currentSolution() {
        return currentSolution;
    }

    @Override
    public ArrayList<SolutionSummary> searchHistory() {
        return searchHistory;
    }

    @Override
    public void stop() {
        stopFlag = true;
    }

    private class Solver extends RecursiveAction {
        public OptimizationAlgorithm solver;
        private  Class optimizerType;
        private OptimizationParameters parameters;
        private Solution startSolution;
        private ArrayList<Solution> outputSolutions;

        public Solver(Class optimizerType, OptimizationParameters parameters, Solution startSolution, ArrayList<Solution> outputSolutions) {
            this.parameters = parameters;
            this.optimizerType = optimizerType;
            this.startSolution = startSolution;
            this.outputSolutions = outputSolutions;
        }

        @Override
        protected void compute() {
            try {
                solver = (OptimizationAlgorithm)optimizerType.getConstructor(parameters.getClass(), SolutionHandler.class).newInstance(parameters, new SolutionProcessor(outputSolutions));
                solver.minimize(startSolution);
            } catch (NoSuchMethodException | InvocationTargetException | InstantiationException | IllegalAccessException e) {
                e.printStackTrace();
            }
        }
    }

    private class SolutionProcessor implements SolutionHandler {

        private ArrayList<Solution> outputSolutions;

        public SolutionProcessor(ArrayList<Solution> outputSolutions) {
            this.outputSolutions = outputSolutions;
        }

        @Override
        public void process(Solution solution) {
            synchronized (thisLock) {
                outputSolutions.add(solution);
            }
        }
    }
}